Automattic/woocommerce-payments
# Some orders do not have corresponding transaction

Fixes #3440 

#### Changes proposed in this Pull Request

This addresses a bug where the intent info would not get properly attached to a UPE order if the `payment_intent.succeeded` webhook was processed before the UPE redirect happened. In this case, the function would exit early before attaching the intent info. I have addressed this in two ways:

1. Attach the intent info to the order as soon as possible, before doing the redirect.
2. Adjust the redirect logic to remove the part which exited early if the order already had a status of `processing`/`competed`/`on-hold`. Because there is a mutex lock on the `payment_complete` logic for an order, it should be safe to do this because there's no danger of the order being marked as completed twice.









#### Testing instructions

Because this bug is intermittent and only seems to affect certain sites, it is hard to test in normal conditions. It is unclear if there are perhaps certain plugin configurations that would cause it to be more prevalent. Because of this, I will use a sleep to test it.

* Enable UPE checkout
* Run the listener with `npm run listen` on your local server.
* Add a `sleep(30)` to the beginning of the `process_redirect_payment` function (line 538 of `class-upe-payment-gateway.php`)
* Make an order using a 3D Secure card: `4000000000003220`
* Wait for the order confirmation page to appear. Refresh the page a few times to make sure the notes don't get duplicated.
* Check the logs, you should see the following: `UPE order <order_id> had status processing, completed or on-hold when UPE redirect processing started`
* Check the order in WC Admin, verify it was created properly, the notes look right (we should see the `A payment of $15.00 USD was successfully charged using WooCommerce Payments (ch_3KH4WfQvx55eNAWf1VwricS7).` note, and no duplication of notes.
* Verify you can see the "Refund using WooCommerce Payments" button after clicking on the "Refund" button. You don't need to actually process a refund, just verifying the button is present is sufficient.
* Go to Payments > Transactions, verify you can see the transaction and it corresponds to the order (has the order ID present).





*

-------------------


**Post merge**



- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : [Link](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-3.6.0#fix-for-upe-checkout-issue)
